onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -2108,7 +2109,87 @@ def aten_convolution_backward(
+    # Assume: grad_output=(20,13,48,38)
+    z_height = op.Shape(grad_output, start=2, end=3)  # 48
+    z_width = op.Shape(grad_output, start=3, end=4)  # 38
+
+    if stride[0] != 1 or stride[1] != 1:
+        raise NotImplementedError("stride != 1 is not supported yet")
+    # if stride[0] != 1:  # dilation
+    #     dz_height = z_height * stride[0] - stride[0] + 1
+    #     dz_width = z_width * stride[1] - stride[1] + 1
+    #     pos = _help(z_height, dz_width, stride)
+    #     pos = []
+    #     for j in range(z_height):
+    #         for i in range(0, dz_width, stride[1]):
+    #             pos.append(i + j * dz_width * stride[0])
+
+    #     index_tensor = op.Constant(value_ints=pos)
+    #     index_tensor = op.Reshape(index_tensor, z_shape)
+    #     # this should not work because the kernel_shape is attribute
+    #     dz = op.MaxUnpool(grad_output, index_tensor, kernel_shape=[dz_height - z_height + 1, dz_width - z_width + 1])
+
+    # # Computing padding size
+    # Assume: input=(20,16,50,40)
+    x_height = op.Shape(input, start=2, end=3)  # 50
+    x_width = op.Shape(input, start=3, end=4)  # 40
+    # Assume: weight=(13,16,3,3)
+    w_height = op.Shape(weight, start=2, end=3)  # 3
+    w_width = op.Shape(weight, start=3, end=4)  # 3
+    tmp_int = x_height - z_height + w_height - 1  # 50-48+3-1=4
+    tmp_float = op.Cast(tmp_int, to=FLOAT.dtype)
+    pad_height = op.Cast(
+        op.Div(tmp_float, op.Constant(value_floats=[2.0])), to=INT64.dtype
+    )  # 4/2=2
+    tmp_int = x_width - z_width + w_width - 1  # 40-38+3-1=4
+    tmp_float = op.Cast(tmp_int, to=FLOAT.dtype)
+    pad_width = op.Cast(
+        op.Div(tmp_float, op.Constant(value_floats=[2.0])), to=INT64.dtype
+    )  # 4/2=2
+    pads = op.Concat(  # [0,0,2,2,0,0,2,2]
+        # begin of dim0, dim1, dim2, dim3
+        op.Constant(value_ints=[0]),
+        op.Constant(value_ints=[0]),
+        pad_height,
+        pad_width,
+        # end of dim0, dim1, dim2, dim3
+        op.Constant(value_ints=[0]),
+        op.Constant(value_ints=[0]),
+        pad_height,
+        pad_width,
+        axis=0,
+    )
+    dz_pad = op.Pad(grad_output, pads)  # enlarge the grad_output to (20,13,52,42)
+
+    # Transpose from (13,16,3,3) to (16,13,3,3)
+    w_transpose = op.Transpose(weight, perm=[1, 0, 2, 3])
+    # Rotate weight (13,16,3,3) with 180 degree: np.rot90(w, 2) -> (13,6,3,3)
+    w_shape_0 = op.Shape(w_transpose, start=0, end=1)  # 13
+    w_shape_1 = op.Shape(w_transpose, start=1, end=2)  # 6
+    w_shape_2 = op.Constant(value_ints=[1])  # 1
+    w_shape_3 = op.Constant(value_ints=[-1])  # -1
+    w_shape_new = op.Concat(w_shape_0, w_shape_1, w_shape_2, w_shape_3, axis=0)  # (13,16,1,-1)
+    w_new = op.Reshape(w_transpose, w_shape_new)  # reshape to (13,16,1,-1)
+    # reverse the values in the last dim (axes=3), e.g. [1,2,3....,9] -> [9,...,3,2,1]
+    starts = op.Constant(value_int=[-1])
+    ends = op.Constant(value_int=[-1000])
```

Comment:
Why 1000?

Reply:
We can compute the element count in a tensor: dim1 * dim2 * dim3... but we will get a tensor which cannot be used in `value_int=[]` attribute. So we have to use a scalar here.
This is a kernel of conv, the most used kernel weight is not more than 11x11, so the -1000 will be large enough get all the data. For example, a kernel shape with 64x128x3x3, will reshape to 64x128x1x9, then slice as [:,:,:,-1:-1000:-1]. I don't know other method to get the whole data.
```
>>> a = np.arange(120).reshape(2,3,4,5)
>>> a
array([[[[  0,   1,   2,   3,   4],
         [ 10,  11,  12,  13,  14],
         [ 15,  16,  17,  18,  19]],

         [ 25,  26,  27,  28,  29],

        [[ 40,  41,  42,  43,  44],
         [ 45,  46,  47,  48,  49],
         [ 50,  51,  52,  53,  54],
         [ 55,  56,  57,  58,  59]]],


       [[[ 60,  61,  62,  63,  64],
         [ 65,  66,  67,  68,  69],
         [ 70,  71,  72,  73,  74],
         [ 75,  76,  77,  78,  79]],

        [[ 80,  81,  82,  83,  84],
         [ 85,  86,  87,  88,  89],
         [ 90,  91,  92,  93,  94],
         [ 95,  96,  97,  98,  99]],

        [[100, 101, 102, 103, 104],
         [115, 116, 117, 118, 119]]]])
>>> b = a.reshape(2,3,1,-1)
>>> b
array([[[[  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,
           12,  13,  14,  15,  16,  17,  18,  19]],

        [[ 20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,
           32,  33,  34,  35,  36,  37,  38,  39]],

        [[ 40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
           52,  53,  54,  55,  56,  57,  58,  59]]],


       [[[ 60,  61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,
           72,  73,  74,  75,  76,  77,  78,  79]],

        [[ 80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
           92,  93,  94,  95,  96,  97,  98,  99]],
          112, 113, 114, 115, 116, 117, 118, 119]]]])

>>> c = b[:,:,:,-1:-1000:-1]
>>> c
array([[[[ 19,  18,  17,  16,  15,  14,  13,  12,  11,  10,   9,   8,
            7,   6,   5,   4,   3,   2,   1,   0]],

        [[ 39,  38,  37,  36,  35,  34,  33,  32,  31,  30,  29,  28,
           27,  26,  25,  24,  23,  22,  21,  20]],


       [[[ 79,  78,  77,  76,  75,  74,  73,  72,  71,  70,  69,  68,
           67,  66,  65,  64,  63,  62,  61,  60]],

        [[ 99,  98,  97,  96,  95,  94,  93,  92,  91,  90,  89,  88,
           87,  86,  85,  84,  83,  82,  81,  80]],

        [[119, 118, 117, 116, 115, 114, 113, 112, 111, 110, 109, 108,
          107, 106, 105, 104, 103, 102, 101, 100]]]])

```